# Exploración Inicial del Dataset

**Dataset:** `student_career_success_dataset.csv`

**Objetivo de esta etapa:** conocer la estructura y la calidad de los datos antes de tomar decisiones de modelado.

In [1]:
import sys

print("Versión de Python:", sys.version)
print("Ruta de Python:", sys.executable)

Versión de Python: 3.12.12 | packaged by conda-forge | (main, Oct 22 2025, 23:34:53) [Clang 19.1.7 ]
Ruta de Python: /Users/juanpablohernandezpizarro/iCloud Drive (archivado)/Documents/Documentos - MacBook Pro de Juan (2)/Respaldo JPHP 2024(345,79GB)/Nombres/Juan Pablo Hernández Pizarro/Magister en Data Science/07_Cloud Computing para Data Science_Ahmad Armoush/Tarea_Final_Cloud_Computing/.venv/bin/python


In [2]:
import pandas as pd

print("Versión de pandas:", pd.__version__)

Versión de pandas: 2.2.3


In [3]:
from pathlib import Path

print("Carpeta de trabajo del notebook:", Path.cwd())

Carpeta de trabajo del notebook: /Users/juanpablohernandezpizarro/iCloud Drive (archivado)/Documents/Documentos - MacBook Pro de Juan (2)/Respaldo JPHP 2024(345,79GB)/Nombres/Juan Pablo Hernández Pizarro/Magister en Data Science/07_Cloud Computing para Data Science_Ahmad Armoush/Tarea_Final_Cloud_Computing/notebooks


In [4]:
ruta_csv = Path("..") / "student_career_success_dataset.csv"

print("Ruta completa:", ruta_csv.resolve())
print("¿Existe el archivo?:", ruta_csv.exists())

Ruta completa: /Users/juanpablohernandezpizarro/iCloud Drive (archivado)/Documents/Documentos - MacBook Pro de Juan (2)/Respaldo JPHP 2024(345,79GB)/Nombres/Juan Pablo Hernández Pizarro/Magister en Data Science/07_Cloud Computing para Data Science_Ahmad Armoush/Tarea_Final_Cloud_Computing/student_career_success_dataset.csv
¿Existe el archivo?: True


In [5]:
df = pd.read_csv(ruta_csv)

print("Dataset cargado correctamente.")
print("Número de filas:", df.shape[0])
print("Número de columnas:", df.shape[1])

Dataset cargado correctamente.
Número de filas: 50000
Número de columnas: 29


In [6]:
df.head(3)

,Student_ID,Age,Gender,University_Year,Major,Attendance_Percentage,Study_Hours_Per_Week,CGPA,Academic_Performance,Programming_Skill,...,Teamwork,Problem_Solving,English_Proficiency,Interview_Score,Employability_Score,Placement_Status,Company_Tier,Career_Field,Placement_Mode,Starting_Salary_USD
0,ST000001,22,Male,Sophomore,Computer Science,86,21,3.28,Good,9,...,7,8,Advanced,72,231.70,Placed,Tier 1,Backend Developer,Campus Placement,72082
1,ST000002,20,Female,Freshman,Computer Science,64,18,2.53,Poor,6,...,6,5,Basic,59,157.45,Not Placed,No Company,Not Placed,Not Applicable,0
2,ST000003,23,Male,Senior,Information Technology,75,19,2.83,Average,8,...,8,7,Advanced,76,203.45,Placed,Tier 1,IT Support Engineer,Campus Placement,94898


In [7]:
df.columns.tolist()

['Student_ID',
 'Age',
 'Gender',
 'University_Year',
 'Major',
 'Attendance_Percentage',
 'Study_Hours_Per_Week',
 'CGPA',
 'Academic_Performance',
 'Programming_Skill',
 'Projects_Completed',
 'Certifications',
 'Hackathons',
 'GitHub_Profile',
 'Internships',
 'Leadership_Experience',
 'LinkedIn_Profile',
 'Resume_Score',
 'Communication_Skills',
 'Teamwork',
 'Problem_Solving',
 'English_Proficiency',
 'Interview_Score',
 'Employability_Score',
 'Placement_Status',
 'Company_Tier',
 'Career_Field',
 'Placement_Mode',
 'Starting_Salary_USD']

In [8]:
df["Placement_Status"].value_counts(dropna=False)

Placement_Status
Placed        39040
Not Placed    10960
Name: count, dtype: int64

In [9]:
df["Placement_Status"].value_counts(
    normalize=True,
    dropna=False
).mul(100).round(2)

Placement_Status
Placed        78.08
Not Placed    21.92
Name: proportion, dtype: float64

## Variable objetivo

Se selecciona `Placement_Status` como variable objetivo porque representa el resultado laboral que se desea predecir.

El problema corresponde a una clasificación supervisada binaria, con dos clases:

- `Placed`: 39.040 estudiantes (78,08 %).
- `Not Placed`: 10.960 estudiantes (21,92 %).

La distribución evidencia un desbalance de clases relevante. Por este motivo, la exactitud (`accuracy`) no se interpretará de manera aislada y deberá complementarse con métricas que evalúen el desempeño sobre ambas clases.

El momento de predicción se define después de la entrevista y antes de conocer el resultado de contratación.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 29 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Student_ID             50000 non-null  object 
 1   Age                    50000 non-null  int64  
 2   Gender                 50000 non-null  object 
 3   University_Year        50000 non-null  object 
 4   Major                  50000 non-null  object 
 5   Attendance_Percentage  50000 non-null  int64  
 6   Study_Hours_Per_Week   50000 non-null  int64  
 7   CGPA                   50000 non-null  float64
 8   Academic_Performance   50000 non-null  object 
 9   Programming_Skill      50000 non-null  int64  
 10  Projects_Completed     50000 non-null  int64  
 11  Certifications         50000 non-null  int64  
 12  Hackathons             50000 non-null  int64  
 13  GitHub_Profile         50000 non-null  object 
 14  Internships            50000 non-null  int64  
 15  Le

In [11]:
total_nulos = df.isna().sum().sum()

print("Total de valores nulos:", total_nulos)

Total de valores nulos: 0


In [12]:
duplicados_exactos = df.duplicated().sum()

print("Filas duplicadas exactas:", duplicados_exactos)

Filas duplicadas exactas: 0


In [13]:
print("Cantidad de IDs únicos:", df["Student_ID"].nunique())
print("Cantidad de IDs repetidos:", df["Student_ID"].duplicated().sum())

Cantidad de IDs únicos: 50000
Cantidad de IDs repetidos: 0


## Calidad inicial de los datos

- El dataset contiene 50.000 filas y 29 columnas.
- No se detectaron valores nulos.
- No se detectaron filas completamente duplicadas.
- `Student_ID` contiene 50.000 valores únicos y 0 identificadores repetidos.
- Cada fila representa un estudiante distinto.

`Student_ID` se conservará en el dataset original, pero se excluirá de las variables predictoras porque es un identificador único y no una característica generalizable para estudiantes futuros.

In [14]:
columnas_object = df.select_dtypes(include="object").columns

df[columnas_object].nunique().sort_values()

GitHub_Profile               2
Leadership_Experience        2
LinkedIn_Profile             2
Placement_Status             2
Gender                       3
English_Proficiency          3
University_Year              4
Academic_Performance         4
Company_Tier                 4
Placement_Mode               5
Major                        8
Career_Field                17
Student_ID               50000
dtype: int64

In [15]:
for columna in columnas_object:
    if columna != "Student_ID":
        valores = sorted(df[columna].unique())
        print(f"{columna}: {valores}")

Gender: ['Female', 'Male', 'Other']
University_Year: ['Freshman', 'Junior', 'Senior', 'Sophomore']
Major: ['Artificial Intelligence', 'Business Analytics', 'Computer Science', 'Cybersecurity', 'Data Science', 'Electrical Engineering', 'Information Technology', 'Software Engineering']
Academic_Performance: ['Average', 'Excellent', 'Good', 'Poor']
GitHub_Profile: ['No', 'Yes']
Leadership_Experience: ['No', 'Yes']
LinkedIn_Profile: ['No', 'Yes']
English_Proficiency: ['Advanced', 'Basic', 'Intermediate']
Placement_Status: ['Not Placed', 'Placed']
Company_Tier: ['No Company', 'Tier 1', 'Tier 2', 'Tier 3']
Career_Field: ['AI Engineer', 'Automation Engineer', 'Backend Developer', 'Business Analyst', 'Cloud Engineer', 'Cybersecurity Analyst', 'Data Analyst', 'Data Scientist', 'DevOps Engineer', 'Embedded Systems Engineer', 'Full Stack Developer', 'IT Support Engineer', 'Machine Learning Engineer', 'Not Placed', 'Penetration Tester', 'SOC Analyst', 'Software Engineer']
Placement_Mode: ['Campus 

In [16]:
pd.crosstab(
    df["Placement_Status"],
    df["Company_Tier"],
    margins=True
)

Company_Tier,No Company,Tier 1,Tier 2,Tier 3,All
Placement_Status,,,,,
Not Placed,10960,0,0,0,10960
Placed,0,37152,1751,137,39040
All,10960,37152,1751,137,50000


In [17]:
pd.crosstab(
    df["Placement_Status"],
    df["Career_Field"].eq("Not Placed"),
    rownames=["Placement_Status"],
    colnames=['Career_Field es "Not Placed"']
)

"Career_Field es ""Not Placed""",False,True
Placement_Status,,
Not Placed,0,10960
Placed,39040,0


In [18]:
pd.crosstab(
    df["Placement_Status"],
    df["Placement_Mode"].eq("Not Applicable"),
    rownames=["Placement_Status"],
    colnames=['Placement_Mode es "Not Applicable"']
)

"Placement_Mode es ""Not Applicable""",False,True
Placement_Status,,
Not Placed,0,10960
Placed,39040,0


In [19]:
df.groupby("Placement_Status")["Starting_Salary_USD"].agg(
    ["count", "min", "max", "mean", "median", "nunique"]
).round(2)

,count,min,max,mean,median,nunique
Placement_Status,,,,,,
Not Placed,10960,0,0,0.00,0.0,1
Placed,39040,25096,110000,88355.75,89025.0,26007


## Variables con fuga de información

El momento de predicción se define después de la entrevista y antes de conocer el resultado de contratación.

Se identificaron cuatro variables generadas o conocidas después del resultado laboral:

- `Company_Tier`: todos los estudiantes `Not Placed` tienen `No Company`, mientras que los estudiantes `Placed` tienen algún nivel de empresa.
- `Career_Field`: la categoría `Not Placed` coincide exactamente con la clase `Not Placed`.
- `Placement_Mode`: la categoría `Not Applicable` coincide exactamente con la clase `Not Placed`.
- `Starting_Salary_USD`: todos los estudiantes `Not Placed` tienen salario 0, mientras que todos los estudiantes `Placed` tienen salario positivo.

Estas variables permiten reconstruir directamente `Placement_Status`, por lo que producirían fuga de información y métricas artificialmente altas.

Se conservarán en el dataset original para mantener su integridad, pero se excluirán de las variables predictoras utilizadas para entrenar el modelo.

In [20]:
df.groupby("Placement_Status")["Employability_Score"].agg(
    ["count", "min", "max", "mean", "median", "std", "nunique"]
).round(3)

,count,min,max,mean,median,std,nunique
Placement_Status,,,,,,,
Not Placed,10960,82.3,274.80,190.061,191.4,27.242,2472
Placed,39040,111.6,279.05,220.709,223.1,22.524,2423


In [21]:
scores_not_placed = set(
    df.loc[df["Placement_Status"] == "Not Placed", "Employability_Score"]
)

scores_placed = set(
    df.loc[df["Placement_Status"] == "Placed", "Employability_Score"]
)

scores_compartidos = scores_not_placed.intersection(scores_placed)

print("Cantidad de puntuaciones compartidas:", len(scores_compartidos))

Cantidad de puntuaciones compartidas: 2003


In [22]:
correlaciones_employability = (
    df.select_dtypes(include="number")
      .corr()["Employability_Score"]
      .drop("Employability_Score")
      .sort_values(ascending=False)
)

correlaciones_employability

Interview_Score          0.844578
Programming_Skill        0.814310
Resume_Score             0.790519
Internships              0.789657
Problem_Solving          0.764074
Projects_Completed       0.752465
CGPA                     0.601362
Communication_Skills     0.556664
Starting_Salary_USD      0.516361
Attendance_Percentage    0.403592
Hackathons               0.364893
Study_Hours_Per_Week     0.317465
Teamwork                 0.300940
Certifications           0.194933
Age                      0.006908
Name: Employability_Score, dtype: float64

## Revisión de `Employability_Score`

`Employability_Score` no reproduce directamente la variable objetivo:

- Los rangos de `Placed` y `Not Placed` se superponen.
- Existen 2.003 valores exactos compartidos entre ambas clases.
- La puntuación presenta correlaciones fuertes con variables disponibles antes de la contratación, especialmente `Interview_Score`, `Programming_Skill`, `Resume_Score`, `Internships` y `Problem_Solving`.

Se conservará provisionalmente como predictor, bajo el supuesto explícito de que esta puntuación está disponible después de la entrevista y antes del resultado laboral.

Como control de sensibilidad, posteriormente se comparará el desempeño del modelo con y sin `Employability_Score`.

In [23]:
df.groupby("Academic_Performance")["CGPA"].agg(
    ["count", "min", "max", "mean", "median"]
).sort_values("min")

,count,min,max,mean,median
Academic_Performance,,,,,
Poor,1956,2.0,2.59,2.478993,2.51
Average,28890,2.6,3.19,2.956083,2.98
Good,18168,3.2,3.69,3.374752,3.35
Excellent,986,3.7,4.00,3.792606,3.77


In [24]:
df["CGPA"].agg(["min", "max", "mean", "median"])

min       2.000000
max       4.000000
mean      3.106043
median    3.110000
Name: CGPA, dtype: float64

## Revisión de `Academic_Performance`

CGPA: Cumulative Grade Point Average (Promedio acumulado de calificaciones de un estudiante durante su carrera)

Las categorías de `Academic_Performance` corresponden exactamente a intervalos de `CGPA`:

- `Poor`: 2,00 a 2,59.
- `Average`: 2,60 a 3,19.
- `Good`: 3,20 a 3,69.
- `Excellent`: 3,70 a 4,00.

`Academic_Performance` es una transformación categórica derivada de `CGPA`.

Se conservará `CGPA` como predictor porque mantiene el valor exacto y se excluirá `Academic_Performance` para evitar información redundante. La columna permanecerá intacta en el dataset original.

In [25]:
target = "Placement_Status"

columnas_excluidas = [
    "Student_ID",
    "Academic_Performance",
    "Company_Tier",
    "Career_Field",
    "Placement_Mode",
    "Starting_Salary_USD",
]

variables_predictoras = [
    columna
    for columna in df.columns
    if columna not in columnas_excluidas + [target]
]

print("Variable objetivo:", target)
print("Columnas excluidas:", len(columnas_excluidas))
print("Variables predictoras:", len(variables_predictoras))
print()
print(variables_predictoras)

Variable objetivo: Placement_Status
Columnas excluidas: 6
Variables predictoras: 22

['Age', 'Gender', 'University_Year', 'Major', 'Attendance_Percentage', 'Study_Hours_Per_Week', 'CGPA', 'Programming_Skill', 'Projects_Completed', 'Certifications', 'Hackathons', 'GitHub_Profile', 'Internships', 'Leadership_Experience', 'LinkedIn_Profile', 'Resume_Score', 'Communication_Skills', 'Teamwork', 'Problem_Solving', 'English_Proficiency', 'Interview_Score', 'Employability_Score']


In [26]:
X = df[variables_predictoras]
y = df[target]

columnas_categoricas = X.select_dtypes(include="object").columns.tolist()
columnas_numericas = X.select_dtypes(include="number").columns.tolist()

print("Forma de X:", X.shape)
print("Forma de y:", y.shape)
print("Predictoras categóricas:", len(columnas_categoricas))
print(columnas_categoricas)
print()
print("Predictoras numéricas:", len(columnas_numericas))
print(columnas_numericas)

Forma de X: (50000, 22)
Forma de y: (50000,)
Predictoras categóricas: 7
['Gender', 'University_Year', 'Major', 'GitHub_Profile', 'Leadership_Experience', 'LinkedIn_Profile', 'English_Proficiency']

Predictoras numéricas: 15
['Age', 'Attendance_Percentage', 'Study_Hours_Per_Week', 'CGPA', 'Programming_Skill', 'Projects_Completed', 'Certifications', 'Hackathons', 'Internships', 'Resume_Score', 'Communication_Skills', 'Teamwork', 'Problem_Solving', 'Interview_Score', 'Employability_Score']


In [27]:
X[columnas_numericas].describe().T

,count,mean,std,min,25%,50%,75%,max
Age,50000.0,21.262900,2.041163,18.0,20.00,21.00,22.00,30.00
Attendance_Percentage,50000.0,81.409640,9.709519,50.0,75.00,82.00,88.00,100.00
Study_Hours_Per_Week,50000.0,21.528160,6.974810,5.0,17.00,21.00,26.00,45.00
CGPA,50000.0,3.106043,0.290351,2.0,2.91,3.11,3.31,4.00
Programming_Skill,50000.0,7.137160,1.507194,1.0,6.00,7.00,8.00,10.00
Projects_Completed,50000.0,8.620560,2.505408,0.0,7.00,9.00,10.00,15.00
Certifications,50000.0,1.715660,1.368301,0.0,1.00,2.00,3.00,8.00
Hackathons,50000.0,3.093940,1.660774,0.0,2.00,3.00,4.00,10.00
Internships,50000.0,4.048880,1.005719,0.0,3.00,4.00,5.00,5.00
Resume_Score,50000.0,96.045740,7.331565,44.0,95.00,100.00,100.00,100.00


## Revisión de variables numéricas

Las variables numéricas utilizan escalas diferentes: `CGPA` varía entre 2 y 4, varias habilidades entre 1 y 10, y otras puntuaciones alcanzan valores cercanos o superiores a 100.

No se detectaron valores nulos ni valores evidentemente imposibles según los rangos observados. `Resume_Score` presenta un efecto techo, ya que su mediana y tercer cuartil son 100; se conservará sin modificaciones porque no existe evidencia suficiente para eliminar o recortar observaciones.

No se aplicará escalamiento directamente sobre el dataset completo. Cualquier transformación numérica se incorporará dentro del pipeline y se ajustará utilizando solamente los datos de entrenamiento, evitando fuga de información.

In [28]:
import sklearn
import joblib

print("Versión de scikit-learn:", sklearn.__version__)
print("Versión de joblib:", joblib.__version__)

Versión de scikit-learn: 1.5.2
Versión de joblib: 1.4.2


In [29]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y,
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (40000, 22)
X_test: (10000, 22)
y_train: (40000,)
y_test: (10000,)


In [30]:
distribucion_split = pd.DataFrame({
    "Entrenamiento (%)": (
        y_train.value_counts(normalize=True).mul(100).round(2)
    ),
    "Prueba (%)": (
        y_test.value_counts(normalize=True).mul(100).round(2)
    ),
})

distribucion_split

,Entrenamiento (%),Prueba (%)
Placement_Status,,
Placed,78.08,78.08
Not Placed,21.92,21.92


## Separación entre entrenamiento y prueba

Se utilizó una división estratificada de los datos:

- 80 % para entrenamiento: 40.000 estudiantes.
- 20 % para prueba: 10.000 estudiantes.
- Semilla fija: `random_state=42`.
- Estratificación mediante `stratify=y`.

La proporción de clases se conservó en ambos conjuntos:

- `Placed`: 78,08 %.
- `Not Placed`: 21,92 %.

El conjunto de prueba se mantendrá aislado durante el ajuste del preprocesamiento y la selección del modelo. Se utilizará únicamente para la evaluación final.

In [31]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

preprocesador = ColumnTransformer(
    transformers=[
        (
            "numericas",
            StandardScaler(),
            columnas_numericas,
        ),
        (
            "categoricas",
            OneHotEncoder(handle_unknown="ignore"),
            columnas_categoricas,
        ),
    ],
    remainder="drop",
)

preprocesador

ColumnTransformer(transformers=[('numericas', StandardScaler(),
                                 ['Age', 'Attendance_Percentage',
                                  'Study_Hours_Per_Week', 'CGPA',
                                  'Programming_Skill', 'Projects_Completed',
                                  'Certifications', 'Hackathons', 'Internships',
                                  'Resume_Score', 'Communication_Skills',
                                  'Teamwork', 'Problem_Solving',
                                  'Interview_Score', 'Employability_Score']),
                                ('categoricas',
                                 OneHotEncoder(handle_unknown='ignore'),
                                 ['Gender', 'University_Year', 'Major',
                                  'GitHub_Profile', 'Leadership_Experience',
                                  'LinkedIn_Profile', 'English_Proficiency'])])

In [32]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

clasificador = LogisticRegression(
    max_iter=1000,
    class_weight="balanced",
    random_state=42,
)

pipeline_modelo = Pipeline(
    steps=[
        ("preprocesamiento", preprocesador),
        ("clasificador", clasificador),
    ]
)

pipeline_modelo

Pipeline(steps=[('preprocesamiento',
                 ColumnTransformer(transformers=[('numericas', StandardScaler(),
                                                  ['Age',
                                                   'Attendance_Percentage',
                                                   'Study_Hours_Per_Week',
                                                   'CGPA', 'Programming_Skill',
                                                   'Projects_Completed',
                                                   'Certifications',
                                                   'Hackathons', 'Internships',
                                                   'Resume_Score',
                                                   'Communication_Skills',
                                                   'Teamwork',
                                                   'Problem_Solving',
                                                   'Interview_Score',
                                                   'Employability_Score']),
                                                 ('categoricas',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Gender', 'University_Year',
                                                   'Major', 'GitHub_Profile',
                                                   'Leadership_Experience',
                                                   'LinkedIn_Profile',
                                                   'English_Proficiency'])])),
                ('clasificador',
                 LogisticRegression(class_weight='balanced', max_iter=1000,
                                    random_state=42))])

In [33]:
from sklearn.model_selection import StratifiedKFold

validacion_cruzada = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42,
)

validacion_cruzada

StratifiedKFold(n_splits=5, random_state=42, shuffle=True)

In [34]:
metricas_cv = {
    "accuracy": "accuracy",
    "balanced_accuracy": "balanced_accuracy",
    "f1_macro": "f1_macro",
    "roc_auc": "roc_auc",
}

metricas_cv

{'accuracy': 'accuracy',
 'balanced_accuracy': 'balanced_accuracy',
 'f1_macro': 'f1_macro',
 'roc_auc': 'roc_auc'}

In [35]:
from sklearn.model_selection import cross_validate

resultados_cv = cross_validate(
    estimator=pipeline_modelo,
    X=X_train,
    y=y_train,
    cv=validacion_cruzada,
    scoring=metricas_cv,
    return_train_score=False,
)

resultados_cv

{'fit_time': array([0.08468318, 0.0847652 , 0.07623291, 0.07774591, 0.09145403]),
 'score_time': array([0.07426381, 0.0746398 , 0.07493711, 0.07476997, 0.07523918]),
 'test_accuracy': array([0.7525  , 0.754125, 0.740625, 0.747625, 0.74775 ]),
 'test_balanced_accuracy': array([0.74555319, 0.74659385, 0.73056791, 0.73170524, 0.7360942 ]),
 'test_f1_macro': array([0.69603519, 0.69750956, 0.68241442, 0.68724813, 0.6891279 ]),
 'test_roc_auc': array([0.8168729 , 0.81686304, 0.8076075 , 0.80639049, 0.80906523])}

In [36]:
resumen_cv = pd.DataFrame({
    metrica: {
        "media": resultados_cv[f"test_{metrica}"].mean(),
        "desviacion_estandar": resultados_cv[f"test_{metrica}"].std(),
    }
    for metrica in metricas_cv
}).T.round(4)

resumen_cv

,media,desviacion_estandar
accuracy,0.7485,0.0047
balanced_accuracy,0.7381,0.0068
f1_macro,0.6905,0.0056
roc_auc,0.8114,0.0046


In [37]:
from sklearn.base import clone

pipeline_sin_balanceo = clone(pipeline_modelo)

pipeline_sin_balanceo.set_params(
    clasificador__class_weight=None
)

resultados_cv_sin_balanceo = cross_validate(
    estimator=pipeline_sin_balanceo,
    X=X_train,
    y=y_train,
    cv=validacion_cruzada,
    scoring=metricas_cv,
    return_train_score=False,
)

resumen_cv_sin_balanceo = pd.DataFrame({
    metrica: {
        "media": resultados_cv_sin_balanceo[f"test_{metrica}"].mean(),
        "desviacion_estandar": resultados_cv_sin_balanceo[f"test_{metrica}"].std(),
    }
    for metrica in metricas_cv
}).T.round(4)

resumen_cv_sin_balanceo

,media,desviacion_estandar
accuracy,0.8195,0.0033
balanced_accuracy,0.6518,0.0046
f1_macro,0.6767,0.0057
roc_auc,0.8114,0.0045


## Selección del balanceo de clases

Se comparó la misma regresión logística con y sin `class_weight="balanced"` mediante validación cruzada estratificada de cinco particiones.

| Métrica | Con balanceo | Sin balanceo |
|---|---:|---:|
| Accuracy | 0,7485 | 0,8195 |
| Balanced accuracy | 0,7381 | 0,6518 |
| F1 macro | 0,6905 | 0,6767 |
| ROC-AUC | 0,8114 | 0,8114 |

La configuración sin balanceo obtiene mayor accuracy porque favorece la clase mayoritaria `Placed`. Sin embargo, reduce de forma importante la balanced accuracy y también disminuye el F1 macro.

Se selecciona `class_weight="balanced"` porque el objetivo es evaluar adecuadamente ambas clases, incluida `Not Placed`, que representa el 21,92 % de los casos. La accuracy se reportará como métrica complementaria, pero no se utilizará de manera aislada.

In [38]:
columnas_numericas_sin_employability = [
    columna
    for columna in columnas_numericas
    if columna != "Employability_Score"
]

preprocesador_sin_employability = ColumnTransformer(
    transformers=[
        (
            "numericas",
            StandardScaler(),
            columnas_numericas_sin_employability,
        ),
        (
            "categoricas",
            OneHotEncoder(handle_unknown="ignore"),
            columnas_categoricas,
        ),
    ],
    remainder="drop",
)

pipeline_sin_employability = Pipeline(
    steps=[
        ("preprocesamiento", preprocesador_sin_employability),
        (
            "clasificador",
            LogisticRegression(
                max_iter=1000,
                class_weight="balanced",
                random_state=42,
            ),
        ),
    ]
)

X_train_sin_employability = X_train.drop(
    columns=["Employability_Score"]
)

resultados_cv_sin_employability = cross_validate(
    estimator=pipeline_sin_employability,
    X=X_train_sin_employability,
    y=y_train,
    cv=validacion_cruzada,
    scoring=metricas_cv,
    return_train_score=False,
)

comparacion_employability = pd.DataFrame({
    "Con Employability_Score": resumen_cv["media"],
    "Sin Employability_Score": [
        resultados_cv_sin_employability[f"test_{metrica}"].mean()
        for metrica in metricas_cv
    ],
}).round(4)

comparacion_employability

,Con Employability_Score,Sin Employability_Score
accuracy,0.7485,0.7485
balanced_accuracy,0.7381,0.7380
f1_macro,0.6905,0.6904
roc_auc,0.8114,0.8114


### Decisión sobre `Employability_Score`

Se comparó el mismo pipeline de regresión logística balanceada mediante
validación cruzada estratificada de 5 folds, usando únicamente el conjunto
de entrenamiento.

Los resultados con y sin `Employability_Score` fueron prácticamente
idénticos. Las diferencias en `balanced_accuracy` y `f1_macro` fueron de
solo 0.0001, mientras que `accuracy` y `roc_auc` no cambiaron.

Por lo tanto, se excluye `Employability_Score` del modelo final porque:

- no aporta una mejora material en las métricas;
- es una puntuación compuesta cuya fórmula no está documentada;
- contiene información redundante respecto de otras variables;
- su exclusión simplifica el contrato de entrada de la futura API.

La columna se conserva intacta en el dataset original, pero no será
utilizada como variable predictora.

In [39]:
# Variables definitivas del modelo
variables_predictoras_finales = [
    columna
    for columna in variables_predictoras
    if columna != "Employability_Score"
]

columnas_numericas_finales = [
    columna
    for columna in columnas_numericas
    if columna != "Employability_Score"
]

columnas_categoricas_finales = columnas_categoricas.copy()

# Conservamos exactamente las mismas filas del split original
X_train_final = X_train[variables_predictoras_finales].copy()
X_test_final = X_test[variables_predictoras_finales].copy()

# Preprocesamiento definitivo
preprocesador_final = ColumnTransformer(
    transformers=[
        (
            "numericas",
            StandardScaler(),
            columnas_numericas_finales,
        ),
        (
            "categoricas",
            OneHotEncoder(handle_unknown="ignore"),
            columnas_categoricas_finales,
        ),
    ],
    remainder="drop",
)

# Pipeline definitivo todavía sin entrenar
modelo_final = Pipeline(
    steps=[
        ("preprocesamiento", preprocesador_final),
        (
            "clasificador",
            LogisticRegression(
                max_iter=1000,
                class_weight="balanced",
                random_state=42,
            ),
        ),
    ]
)

print("Variables predictoras finales:", len(variables_predictoras_finales))
print("Numéricas:", len(columnas_numericas_finales))
print("Categóricas:", len(columnas_categoricas_finales))
print("X_train_final:", X_train_final.shape)
print("X_test_final:", X_test_final.shape)
print("\nVariables ordenadas:")
print(variables_predictoras_finales)

Variables predictoras finales: 21
Numéricas: 14
Categóricas: 7
X_train_final: (40000, 21)
X_test_final: (10000, 21)

Variables ordenadas:
['Age', 'Gender', 'University_Year', 'Major', 'Attendance_Percentage', 'Study_Hours_Per_Week', 'CGPA', 'Programming_Skill', 'Projects_Completed', 'Certifications', 'Hackathons', 'GitHub_Profile', 'Internships', 'Leadership_Experience', 'LinkedIn_Profile', 'Resume_Score', 'Communication_Skills', 'Teamwork', 'Problem_Solving', 'English_Proficiency', 'Interview_Score']


In [40]:
modelo_final.fit(X_train_final, y_train)

print("Pipeline definitivo entrenado correctamente.")
print(
    "Clases aprendidas:",
    modelo_final.named_steps["clasificador"].classes_,
)

Pipeline definitivo entrenado correctamente.
Clases aprendidas: ['Not Placed' 'Placed']


In [41]:
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    f1_score,
    roc_auc_score,
)

# Predicción de clases
y_pred_test = modelo_final.predict(X_test_final)

# Probabilidad correspondiente explícitamente a la clase "Placed"
clases_modelo = list(
    modelo_final.named_steps["clasificador"].classes_
)
indice_placed = clases_modelo.index("Placed")

y_prob_placed_test = modelo_final.predict_proba(
    X_test_final
)[:, indice_placed]

# Target binario únicamente para calcular ROC-AUC
y_test_binario = (y_test == "Placed").astype(int)

metricas_test = {
    "accuracy": accuracy_score(y_test, y_pred_test),
    "balanced_accuracy": balanced_accuracy_score(
        y_test,
        y_pred_test,
    ),
    "f1_macro": f1_score(
        y_test,
        y_pred_test,
        average="macro",
    ),
    "roc_auc": roc_auc_score(
        y_test_binario,
        y_prob_placed_test,
    ),
}

pd.Series(
    metricas_test,
    name="valor_test",
).round(4)

accuracy             0.7414
balanced_accuracy    0.7296
f1_macro             0.6825
roc_auc              0.8031
Name: valor_test, dtype: float64

In [42]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
)

orden_clases = ["Not Placed", "Placed"]

matriz_confusion = pd.DataFrame(
    confusion_matrix(
        y_test,
        y_pred_test,
        labels=orden_clases,
    ),
    index=[
        "Real: Not Placed",
        "Real: Placed",
    ],
    columns=[
        "Predicho: Not Placed",
        "Predicho: Placed",
    ],
)

reporte_clasificacion = pd.DataFrame(
    classification_report(
        y_test,
        y_pred_test,
        labels=orden_clases,
        output_dict=True,
        zero_division=0,
    )
).T.round(4)

display(matriz_confusion)
display(reporte_clasificacion)

,Predicho: Not Placed,Predicho: Placed
Real: Not Placed,1553,639
Real: Placed,1947,5861


,precision,recall,f1-score,support
Not Placed,0.4437,0.7085,0.5457,2192.0000
Placed,0.9017,0.7506,0.8193,7808.0000
accuracy,0.7414,0.7414,0.7414,0.7414
macro avg,0.6727,0.7296,0.6825,10000.0000
weighted avg,0.8013,0.7414,0.7593,10000.0000


### Evaluación final sobre el conjunto de prueba

El pipeline definitivo se evaluó una sola vez sobre 10.000 observaciones
reservadas mediante la separación estratificada inicial.

Resultados:

- Accuracy: 0.7414
- Balanced accuracy: 0.7296
- F1 macro: 0.6825
- ROC-AUC: 0.8031

La accuracy no se utiliza como única métrica porque el 78,08% de las
observaciones pertenece a la clase `Placed`. Un clasificador que predijera
siempre esa clase obtendría una accuracy alta, pero no identificaría ningún
caso `Not Placed`.

El modelo identificó correctamente 1.553 de los 2.192 casos `Not Placed`,
alcanzando un recall de 0.7085 para la clase minoritaria. Para la clase
`Placed`, obtuvo un recall de 0.7506 y una precisión de 0.9017.

El balanceo de clases reduce la accuracy global, pero permite distribuir
mejor la capacidad predictiva entre ambas clases. Las métricas del conjunto
de prueba son cercanas a las obtenidas mediante validación cruzada, lo que
indica una generalización razonablemente estable.

In [43]:
import sys
import numpy as np
import pandas as pd
import scipy
import sklearn
import joblib

versiones_entorno = {
    "python": sys.version.split()[0],
    "numpy": np.__version__,
    "pandas": pd.__version__,
    "scipy": scipy.__version__,
    "scikit-learn": sklearn.__version__,
    "joblib": joblib.__version__,
}

pd.Series(
    versiones_entorno,
    name="version",
)

python          3.12.12
numpy             2.1.3
pandas            2.2.3
scipy            1.14.1
scikit-learn      1.5.2
joblib            1.4.2
Name: version, dtype: object

In [44]:
from pathlib import Path
import json

# Desde notebooks/ retrocedemos a la raíz y entramos en model/
carpeta_modelo = Path("../model")
ruta_modelo = carpeta_modelo / "model.pkl"
ruta_metadata = carpeta_modelo / "metadata.json"

# Guardar el pipeline completo:
# preprocesamiento + regresión logística entrenada
joblib.dump(
    modelo_final,
    ruta_modelo,
)

# Resumen de la validación cruzada del modelo definitivo
metricas_cv_final = {
    metrica: {
        "media": round(
            float(
                resultados_cv_sin_employability[
                    f"test_{metrica}"
                ].mean()
            ),
            4,
        ),
        "desviacion_estandar": round(
            float(
                resultados_cv_sin_employability[
                    f"test_{metrica}"
                ].std()
            ),
            4,
        ),
    }
    for metrica in metricas_cv
}

metadata = {
    "problem_type": "binary_classification",
    "target": "Placement_Status",
    "positive_class": "Placed",
    "classes": [
        str(clase)
        for clase in modelo_final
        .named_steps["clasificador"]
        .classes_
    ],
    "input_features": variables_predictoras_finales,
    "numeric_features": columnas_numericas_finales,
    "categorical_features": columnas_categoricas_finales,
    "excluded_features": [
        "Student_ID",
        "Academic_Performance",
        "Company_Tier",
        "Career_Field",
        "Placement_Mode",
        "Starting_Salary_USD",
        "Employability_Score",
    ],
    "dataset_rows": len(X),
    "training_rows": len(X_train_final),
    "test_rows": len(X_test_final),
    "data_split": {
        "test_size": 0.20,
        "random_state": 42,
        "stratified": True,
    },
    "cross_validation": {
        "method": "StratifiedKFold",
        "n_splits": 5,
        "shuffle": True,
        "random_state": 42,
    },
    "estimator": {
        "name": "LogisticRegression",
        "max_iter": 1000,
        "class_weight": "balanced",
        "random_state": 42,
    },
    "metrics": {
        "cross_validation": metricas_cv_final,
        "test": {
            nombre: round(float(valor), 4)
            for nombre, valor in metricas_test.items()
        },
    },
    "versions": versiones_entorno,
}

with ruta_metadata.open(
    mode="w",
    encoding="utf-8",
) as archivo:
    json.dump(
        metadata,
        archivo,
        ensure_ascii=False,
        indent=2,
    )

tamano_modelo_mb = (
    ruta_modelo.stat().st_size / (1024 ** 2)
)

print("Modelo guardado en:")
print(ruta_modelo.resolve())

print("\nMetadatos guardados en:")
print(ruta_metadata.resolve())

print(
    "\nTamaño de model.pkl:",
    round(tamano_modelo_mb, 4),
    "MB",
)

Modelo guardado en:
/Users/juanpablohernandezpizarro/iCloud Drive (archivado)/Documents/Documentos - MacBook Pro de Juan (2)/Respaldo JPHP 2024(345,79GB)/Nombres/Juan Pablo Hernández Pizarro/Magister en Data Science/07_Cloud Computing para Data Science_Ahmad Armoush/Tarea_Final_Cloud_Computing/model/model.pkl

Metadatos guardados en:
/Users/juanpablohernandezpizarro/iCloud Drive (archivado)/Documents/Documentos - MacBook Pro de Juan (2)/Respaldo JPHP 2024(345,79GB)/Nombres/Juan Pablo Hernández Pizarro/Magister en Data Science/07_Cloud Computing para Data Science_Ahmad Armoush/Tarea_Final_Cloud_Computing/model/metadata.json

Tamaño de model.pkl: 0.0069 MB
